docs/notebook_validation.py

```diff
@@ -26,15 +28,16 @@ def read_available_backends():
 
 def validate(notebook_filename, available_backends):
     with open(notebook_filename) as f:
-        lines = f.readlines()
-    for notebook_content in lines:
-        match = re.search('set_target[\\\s\(]+"(.+)\\\\"[)]', notebook_content)
-        if match and (match.group(1) not in available_backends):
-            return False
-    for notebook_content in lines:
-        match = re.search('--target ([^ ]+)', notebook_content)
-        if match and (match.group(1) not in available_backends):
-            return False
+        lines = f.read()
+    if any(
+            re.search(pattern, lines) for pattern in
+        ['set_target[\\\s\(]+"(.+)\\\\"[)]', '--target ([^ ]+)']):
+        matches = re.findall(
+            'set_target[\\\s\(]+"(.+)\\\\"[)]|--target ([^ ]+)', lines)
```

Comment:
I'm having trouble figuring out what this code change is doing. It looks more confusing to me with the duplicate strings in there. What exactly is this optimizing?

Reply:
> I'm having trouble figuring out what this code change is doing. It looks more confusing to me with the duplicate strings in there. What exactly is this optimizing?

And I'm also curious which notebooks (if any) show a measurable difference in runtime because of this optimization?